src/adapter/src/session.rs

```diff
@@ -273,18 +274,28 @@ impl<T: TimestampManipulation> Session<T> {
                         }
                         *ops = add_ops;
                     }
-                    TransactionOps::Peeks(txn_ts) => match add_ops {
-                        TransactionOps::Peeks(add_ts) => {
-                            match (&txn_ts, add_ts) {
-                                (Some(txn_ts), Some(add_ts)) => assert_eq!(*txn_ts, add_ts),
-                                (None, Some(add_ts)) => *txn_ts = Some(add_ts),
-                                _ => {}
+                    TransactionOps::Peeks(txn_read_context) => match add_ops {
+                        TransactionOps::Peeks(add_read_context) => {
+                            match (&txn_read_context, add_read_context) {
+                                (
+                                    ReadContext::TimelineTimestamp(txn_timeline, txn_ts),
+                                    ReadContext::TimelineTimestamp(add_timeline, add_ts),
+                                ) => {
+                                    assert_eq!(*txn_timeline, add_timeline);
+                                    assert_eq!(*txn_ts, add_ts);
+                                }
+                                (ReadContext::NoTimestamp, add_read_context) => {
+                                    *txn_read_context = add_read_context
+                                }
+                                (_, ReadContext::NoTimestamp) => {}
                             };
                         }
                         // Iff peeks thus far do not have a timestamp (i.e.
                         // they are constant), we can switch to a write
                         // transaction.
-                        writes @ TransactionOps::Writes(..) if txn_ts.is_none() => {
+                        writes @ TransactionOps::Writes(..)
+                            if !txn_read_context.contains_timeline() =>
```

Comment:
```suggestion
                            if !txn_read_context.contains_timestamp() =>
```

This method name matches the comment more closely, but I think has the same result.

Reply:
Yep, just fixed.